lib/Rabbit.js

```diff
@@ -0,0 +1,253 @@
+        host: "localhost",
+        port: "5627",
+        user: "guest",
+        password: "",
+        secure: false,
+        connectionString: "",
+    },
+    serviceName: "unnamed-service",
+    namespace: "default",
+    prefetch: 5,
+    /*
+     * pubPayload?: {},                 // Necessary for: pub + pubQueue
+     * pubQueue?: {                     // Necessary for: pubQueue
+     *    [serviceName]: string[],      // `eventName`
+     * },
+     * pub?: {                          // Necessary for: pub
+     *    [eventName]: string,          // `triggerName`
+     * },
+     * subQueue?: {                     // Necessary for: subQueue
+     *    [eventName]: (payload) => void,
+     * },
+     * publisher?: {                    // Necessary for: sub
+     *    [registryName: string {
+     *        [eventName]: string,
+     *    }
+     * },
+     * sub: {                           // Necessary for: sub
+     *    [registryName: string {
+     *        [eventName]: (payload) => void,
+     *    }
+     * },
+     * subNotification: {               // Necessary for: sub
+     *    [eventName]: (payload) => void,
+     * },
+     */
+};
+
+function getCompleteConfig(config) {
+    const completeConfig = Object.assign({}, defaultConfig, config);
+    return completeConfig;
+}
+
+function getConnectionString(config) {
+    const { secure, user, password, host, port, connectionString } = config.connection;
+    if (connectionString) {
+        return connectionString;
+    } else {
+        const protocol = (typeof secure === "boolean" && secure) ? "amqps" : "amqp";
+        const protocolBasedPort = protocol === "amqps" ? 5671 : 5672;
+        if (user && password) {
+            return `${protocol}://${user}:${password}@${host}:${port}`;
+        }
+        return `${protocol}://${host}:${port}`;
+    }
+}
+
+async function sleep(delay) {
+    return new Promise((resolve, reject) => {
+        setTimeout(resolve, delay);
+    });
+}
+
+class Rabbit {
+
+    constructor(config = {}, logger = console) {
+        this._publishedQueueList = []; // queue name that has been published (to avoid redundant publication)
+        this._connection = null;
+        this._config = getCompleteConfig(config);
+        this._logger = logger;
+        this._connectionString = getConnectionString(this._config);
+        this._run();
+    }
+
+    async publish(payload) {
+        if (this._connection) {
+            this._config.pubPayload = payload;
+            this._publishedQueueList = [];
+            await this._pubQueue(this._connection);
+            await this._pub(this._connection);
+        } else {
+            await sleep(1000);
+            await this.publish(payload);
+        }
+    }
+
+    async _run() {
+        try {
+            // create connection
+            const connection = await amqp.connect(this._connectionString);
+            this._logger.info("Connected to rmq.");
+            connection.on("error", async (error) => {
```

Comment:
mas go , kayaknya  mendingan instantiate dulu this._connection baru set event nya

Reply:
Boleh